### PR TITLE
New function for multiple phenotypic filter apply

### PR DESCRIPTION
**Pull request by sk-sahu**
* Adds new function `cb_apply_filter()` and `cb_apply_filter_dry_run()` You can read [this section from README.md](https://github.com/lifebit-ai/cloudos/tree/database-update#apply-and-save-a-filter) to understand the usecase.
* Few files remaining to make more standard naming based on its type of function it contains. 

This PR also have sub PRs- #17  #21  #25 #26 #27 (Description in respective PR)

## Review notes 

Apologises for this long PR, this used as a parent branch for other features. So this is treated as `dev-2` branch. Now onward will try to use `dev` as parent branch. 

But most files here are by products of unit test (.json) or document (.Rd) and some cache (.png), so checking R files should be good for review.

### CODE REVIEW

**Comment by sk-sahu** [comment]
First merge - https://github.com/lifebit-ai/cloudos/pull/17 then this one.

**Review by sk-sahu — COMMENTED** [review]
Self reviewed.

File: R/cb_cohort_extract.R [hunk 1]
```diff
@@ -72,14 +70,15 @@ cb_get_samples_table <- function(cloudos,
...
                               page_size = 10) {
-  # make column json
-  columns <- .get_column_json()
-  # make search json
+
+  # make json body
   if(missing(cohort)){
     search = list()
+    columns = list()
   }else{
```
Comment (cgpu) [inline R/cb_cohort_extract.R:78]: Note here if it can be replaced with a more explicit if statement.

File: R/cb_cohort_extract.R [hunk 2]
```diff
@@ -126,6 +123,9 @@ cb_get_samples_table <- function(cloudos,
   # rename the dataframe with column names
   colnames(res_df_new) <- columns_names
   
+  # rename rownames
+  rownames(res_df_new) <- 1:page_size
```
Comment (cgpu) [inline R/cb_cohort_extract.R:127]: Can we make it illegal and stop with error when a `page_size` == 0?

File: R/cb_cohorts_list.R [hunk 3]
```diff
@@ -21,9 +21,7 @@ cb_list_cohorts <- function(cloudos,
                  query = list("teamId" = cloudos@team_id,
                               "pageNumber" = 0,
                               "pageSize" = size))
-  if (!r$status_code == 200) {
-    stop("No cohorts found. Or not able to connect with server.")
-  }
+  httr::stop_for_status(r, task = "list cohorts")
```
Comment (cgpu) [inline R/cb_cohorts_list.R:24]: Maybe an error that more explicitly points to the function would be helpful in a scenario that a user is not directly running the function but calling it from a wrapper script?
 
```diff
- Error in function `cloudos::cb_cohorts_list()`
```

File: R/cb_filter_apply.R [hunk 4]
```diff
@@ -0,0 +1,197 @@
+# example 
+# for a range filter 
+# .build_filter_body(filter_query = list("22" = list("from" = "2015-05-13", "to" = "2016-04-29")))
+# for a value filter 
+# .build_filter_body(filter_query = list ("50" = c("Father", "Mother")))
+# for multiple filters 
+# .build_filter_body(filter_query = list("22" = list("from" = "2015-05-13", "to" = "2016-04-29"), "50" = c("Father", "Mother")))
```
Comment (cgpu) [inline R/cb_filter_apply.R:7]: Cleanup for comments and migrate commands to vignette.

File: R/cb_filter_explore.R [hunk 5]
```diff
@@ -22,12 +22,14 @@ cb_search_phenotypic_filters <- function(cloudos,
...
                               "term" = term))
-  if (!r$status_code == 200) {
-    stop("Something went wrong.")
-  }
+  httr::stop_for_status(r, task = NULL)
   res <- httr::content(r)
   filters <- res$filters
-  message("Total number of filters - ", length(filters))
+  
+  if(length(filters) == 0) stop(message("No phenotypic filters found with - ", term ))
```
Comment (cgpu) [inline R/cb_filter_explore.R:29]: To check if `stop(message())` is good practice. Eg search in GitHub for other people using it or  StackOverflow.

File: R/cb_json.R [hunk 6]
```diff
@@ -80,42 +80,27 @@
 # Column JOSN
```
Comment (cgpu) [inline R/cb_json.R:80]: Consider adding Roxygen2 comments for internal functions.

File: R/cb_json.R [hunk 7]
```diff
@@ -80,42 +80,27 @@
...
+  cohort_columns = list()
+  # check if at all the cohort have filed or send an empty list
+  if(length(my_cohort$columns) == 0){
+    return(cohort_columns)
+  }
+  
+  for(col in my_cohort$columns){
+    col_temp = list("id" = col$id,
+                    "instance" = col$instance,
+                    "array" = col$array # make sure toJSON(auto_unbox = T)
```
Comment (cgpu) [inline R/cb_json.R:94]: Check where does this comment refer to.

File: R/cb_json.R [hunk 8]
```diff
@@ -80,42 +80,27 @@
 # Column JOSN
 # TODO work on column - At this point NO end-point that returns this information, there are cards
```
Comment (cgpu) [inline R/cb_json.R:81]: Check if this is needed

File: R/cb_json.R [hunk 9]
```diff
@@ -80,42 +80,27 @@
...
+  for(col in my_cohort$columns){
+    col_temp = list("id" = col$id,
+                    "instance" = col$instance,
+                    "array" = col$array # make sure toJSON(auto_unbox = T)
     )
-  return(columns)
+    # create a list for first time and append a new list
+    if(length(cohort_columns) == 0){
+      cohort_columns = list(col_temp)
+    }else{
```
Comment (cgpu) [inline R/cb_json.R:99]: Check if we can be more explicit with if

File: README.Rmd [hunk 10]
```diff
@@ -197,19 +163,124 @@ total_participants_with_filter <- cloudos:::cb_filter_participants(my_cloudos,
  
 total_participants_with_filter
 ```
-#### Save a filter
+
+### Apply and Save a filter
 
 Save a filter into the database.
 
+For this a `filter_query` is required.
```
Comment (cgpu) [inline README.Rmd:171]: @cgpu to test README.md commands.

**Review by cgpu — CHANGES_REQUESTED** [review]
Added comments. Looks good!

Check notes and update changelog to reflect changes.

File: R/cb_filter_explore.R (hunk 5 above)
Comment (sk-sahu) [reply R/cb_filter_explore.R:26]: replaced with `httr::stop_for_status()`

File: R/cb_cohort_extract.R (hunk 1 above)
Comment (sk-sahu) [reply R/cb_cohort_extract.R:112]: This creates empty list objects to hold the values

File: R/cb_cohort_extract.R (hunk 2 above)
Comment (sk-sahu) [reply R/cb_cohort_extract.R:164]: This is a good suggestion. Thanks! Addressed https://github.com/lifebit-ai/cloudos/pull/13/commits/c0b28bbcdc7c34c48ad5b6d4a486baeda4f28594

File: R/cb_filter_apply.R (hunk 4 above)
Comment (sk-sahu) [reply R/cb_filter_apply.R]: cleaned

File: R/cb_json.R (hunk 6 above)
Comment (sk-sahu) [reply R/cb_json.R:85]: https://github.com/lifebit-ai/cloudos/pull/13/commits/f2467d3f56cee1255f7b61e8405217b208e356f8

File: R/cb_json.R (hunk 7 above)
Comment (sk-sahu) [reply R/cb_json.R:99]: This comment refers to same line and I suppose this is allowed!!

File: R/cb_json.R (hunk 8 above)
Comment (sk-sahu) [reply R/cb_json.R:86]: This endpoint suppose to be coming

**Pull request merged**